diff --git a/binding.gyp b/binding.gyp
index bb40aa7e8d41644ba690761e9064fc1d0f2dcd6b..f1bab392e0d14f516813802e1f51282a15a904a3 100644
--- a/binding.gyp
+++ b/binding.gyp
@@ -25,13 +25,13 @@
         "NAPI_VERSION=<(napi_build_version)",
       ],
       "cflags_cc": [
-        "-std=c++17"
+        "-std=<(cxxstd)"
       ],
       "conditions": [
         ["OS=='mac'", {
           "xcode_settings": {
             "GCC_SYMBOLS_PRIVATE_EXTERN": "YES", # -fvisibility=hidden
-            "CLANG_CXX_LANGUAGE_STANDARD": "c++17",
+            "CLANG_CXX_LANGUAGE_STANDARD": "<(cxxstd)",
             "MACOSX_DEPLOYMENT_TARGET": "10.9",
           },
         }],
@@ -39,7 +39,7 @@
           "msvs_settings": {
             "VCCLCompilerTool": {
               "AdditionalOptions": [
-                "/std:c++17",
+                "/std:<(cxxstd)",
               ],
               "RuntimeLibrary": 0,
             },
@@ -76,5 +76,6 @@
     "openssl_fips": "",
     "v8_enable_pointer_compression%": 0,
     "v8_enable_31bit_smis_on_64bit_arch%": 0,
+    "cxxstd%": "<!(node -p \"parseInt(process.env.npm_config_target ?? process.versions.node) < 22 ? 'c++17' : 'c++20'\")",
   }
 }
